File: algorithm/baekjoon/src/main/java/com/keunwon/algorithm/bruteforce/Problem15657.java
package com.keunwon.algorithm.bruteforce;

import java.io.BufferedWriter;
import java.io.IOException;
import java.io.OutputStreamWriter;
import java.util.Arrays;
import java.util.Scanner;

public class Problem15657 {
    public static final BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

    public static int[] numbers;
    public static int[] result;

    public static void main(String[] args) throws IOException {
        Scanner sc = new Scanner(System.in);
        final int N = sc.nextInt();
        final int M = sc.nextInt();

        numbers = new int[N];
        result = new int[M];

        for (int i = 0; i < N; i++) {
            numbers[i] = sc.nextInt();
        }

        Arrays.sort(numbers);
        dfs(N, M, 0 , 0);

        bw.flush();
        bw.close();
    }

    public static void dfs(int n, int m, int start, int depth) throws IOException {
        if (m == depth) {
            for (int i : result) {
                bw.write(i + " ");
            }
            bw.newLine();
            return;
        }

        for (int i = start; i < n; i++) {
            result[depth] = numbers[i];
            dfs(n, m, i, depth + 1);
        }
    }
}

File: algorithm/kotlin-algorithm/src/main/kotlin/com/keunwon/algorithm/baekjoon/Problem15657.kt
package com.keunwon.algorithm.baekjoon

/**
 * Title: N과 M (8)
 * Level: 실버-3
 **/
class Problem15657 {
    private val answer = mutableListOf<String>()

    fun solution(m: Int, numbers: IntArray): List<String> {
        numbers.sort()
        dfs(0, 0, IntArray(m), numbers)
        return answer
    }

    private fun dfs(
        depth: Int,
        startIndex: Int,
        picks: IntArray,
        numbers: IntArray,
    ) {
        if (depth == picks.size) {
            answer.add(picks.joinToString(" "))
            return
        }

        for (i in startIndex until numbers.size) {
            picks[depth] = numbers[i]
            dfs(depth + 1, i, picks, numbers)
        }
    }
}

fun main() = with(System.`in`.bufferedReader()) {
    val bw = System.out.bufferedWriter()

    val (n, m) = readLine().split(" ").map { it.toInt() }
    val numbers = readLine().split(" ").map { it.toInt() }.toIntArray()

    Problem15657().solution(m, numbers).forEach {
        bw.write(it)
        bw.newLine()
    }

    bw.flush()
    bw.close()
    close()
}
